### ISSUE TITLE
[clang] Wuninitialized & Wuninitialized-const-reference - warning does not see implicit casts to base class

### ISSUE THREAD

**Issue by PiotrZSL** [llvm/llvm-project#63066]
I run into this issue in production code, here is example:
```
#include <vector>

struct Base {
    std::vector<int> value;
};

struct Parent : Base {
    Parent(Base&, int);
};


void test(Base& obj) {
    {
        Parent obj(obj, 1);
    }
}
```
https://godbolt.org/z/s1fjEohTT

There are 3 issues here:
- variable shadowing, that's being detected with -Wshadow, 
- injection of self into self (that is stupid because constructor can access `this`
- "access" to uninitialized object

Did some tests:
```
#include <vector>

struct Base {
    std::vector<int> value;
};

struct Parent : Base {    
   Parent();
    Parent(const Parent&);
    Parent(Parent&&);
    Parent(Base, int);
    Base base;
};

struct ParentRef : Base {
    ParentRef();
    ParentRef(const Base&, int);
    Base base;
};

void test() {
    Parent obj1(obj1, 1);
    // detected by clang-analyzer-core.CallAndMessage
    // warning: Passed-by-value struct argument contains uninitialized data (e.g., field: 'value')

    Parent obj2(obj2, 2);
    // not detected (clang-analyzer-core.CallAndMessage limitation to only first issue ?)

    Parent obj3(obj3.base, 3);
    // detected by -Wuninitialized
    // warning: variable 'obj' is uninitialized when used within its own initialization

    Parent obj4(obj4);
    // detected by -Wuninitialized
    // warning: variable 'obj4' is uninitialized when used within its own initialization

    Parent obj5(std::move(obj5));
    // detected by -Wuninitialized
    // warning: variable 'obj5' is uninitialized when used within its own initialization

    ParentRef obj6(obj6, 1);
    // not detected

    ParentRef obj7(obj7, 2);
    // not detected

    ParentRef obj8(obj8.base, 3);
    // not detected
}
```

https://godbolt.org/z/qfbTe3xPc

Overall looks like issue is not detected when conversion to base class happens.
For example here:
```
Parent copy(const Parent& p) { return p; }
Parent obj4(copy(obj4));
```
This being detected as -Wuninitialized-const-reference
When this:

```
ParentRef obj6(obj6, 1);
```
is not detected.

Same with this:
```
Parent obj2(obj2, 2);
```
even that copy were made from uninitialized object, is not detected by -Wuninitialized.
